File: Data-Structures-Algorithms-In-TS/AlgorithmsChapters/DP/LongestCommonSubseq.ts
function LCSubStr(st1: string, st2: string): number {
	const X: string[] = st1.split("");
	const Y: string[] = st2.split("");
	const m: number = st1.length;
	const n: number = st2.length;
	const dp: number[][] = new Array(m + 1);
	const p: number[][] = new Array(m + 1);

	for (let i = 0; i <= m; i++) {
		dp[i] = new Array(n + 1).fill(0);
		p[i] = new Array(n + 1).fill(0);
	}

	for (let i = 1; i <= m; i++) {
		for (let j = 1; j <= n; j++) {
			if (X[i - 1] === Y[j - 1]) {
				dp[i][j] = dp[i - 1][j - 1] + 1;
				p[i][j] = 0;
			} else {
				dp[i][j] = dp[i - 1][j] > dp[i][j - 1] ? dp[i - 1][j] : dp[i][j - 1];
				p[i][j] = dp[i - 1][j] > dp[i][j - 1] ? 1 : 2;
			}
		}
	}

	console.log(PrintLCS(p, X, m, n));
	return dp[m][n];
}

function PrintLCS(p: number[][], X: string[], i: number, j: number): string {
	if (i === 0 || j === 0) return "";
	if (p[i][j] === 0) {
		return PrintLCS(p, X, i - 1, j - 1) + X[i - 1];
	} else if (p[i][j] === 1) {
		return PrintLCS(p, X, i - 1, j);
	} else {
		return PrintLCS(p, X, i, j - 1);
	}
}

const X = "carpenter";
const Y = "sharpener";
console.log(LCSubStr(X, Y));

// arpener
// 7

File: Data-Structures-and-Algorithms-using-Swift/AlgorithmsChapters/DP/LongestCommonSubseq.swift
import Foundation

func LCSubSeq(_ st1 : String, _ st2 : String) -> Int {
	var X : [Character] = Array(st1)
	let Y : [Character] = Array(st2)
	let m : Int = st1.count
	let n : Int = st2.count
	var dp : [[Int]] = Array(repeating: Array(repeating: 0, count: n + 1), count: m + 1)
	var p : [[Int]] = Array(repeating: Array(repeating: 0, count: n + 1), count: m + 1)

	// Fill dp array in bottom up fashion.
	var i : Int = 1
	while (i <= m) {
		var j : Int = 1
		while (j <= n) {
			if (X[i - 1] == Y[j - 1]) {
				dp[i][j] = dp[i - 1][j - 1] + 1
				p[i][j] = 0
			} else {
				dp[i][j] = (dp[i - 1][j] > dp[i][j - 1]) ? dp[i - 1][j] : dp[i][j - 1]
				p[i][j] = (dp[i - 1][j] > dp[i][j - 1]) ? 1 : 2
			}
			j += 1
		}
		i += 1
	}
	
	PrintLCS(&p, &X,m,n)
	print()
	return dp[m][n]
}

func PrintLCS(_ p : inout [[Int]], _ X : inout [Character], _ i : Int, _ j : Int) {
	if (i == 0 || j == 0) {
		return
	}
	if (p[i][j] == 0) {
		PrintLCS( &p, &X,i - 1,j - 1)
		print(X[i - 1],terminator: "")
	} else if (p[i][j] == 1) {
		PrintLCS( &p, &X,i - 1,j)
	} else {
		PrintLCS( &p, &X,i,j - 1)
	}
}

// Testing code.
let X : String = "carpenter"
let Y : String = "sharpener"
print(LCSubSeq(X,Y))

/* 
arpener
7
 */